src/main/java/baritone/behavior/PathingBehavior.java

```diff
@@ -565,4 +577,66 @@ private static AbstractNodeCostSearch createPathfinder(BlockPos start, Goal goal
     public void onRenderPass(RenderEvent event) {
         PathRenderer.render(event, this);
     }
+
+    @Override
+    public boolean isCloseToHome() {
+        return isCloseToHome;
+    }
+
+    @Override
+    public void updateIsCloseToHome() {
+        try {
+            EntityPlayerSP player = baritone.getPlayerContext().player();
+            IWaypoint[] waypoints = ForWaypoints.getWaypointsByTag(baritone, IWaypoint.Tag.HOME);
+            if (waypoints.length > 0) {
+                for (IWaypoint wp : waypoints) {
+                    boolean isClose = wp.getLocation().getDistance((int) Math.round(player.posX), (int) Math.round(player.posY), (int) Math.round(player.posZ)) <= Baritone.settings().homeProtectionRange.value;
+                    if (isClose) {
+                        isCloseToHome = true;
+                        logProtectedHomeArea();
+                        return;
+                    }
+                }
+            }
+            isCloseToHome = false;
+            logProtectedHomeArea();
+            return;
+        }
+        catch (Exception err) {
+            return;
+        }
+    }
+
+    //FIXME: The waypoint list gets refreshed every single time CalculationContext or BuilderProcess wants to check if a block is within protected home area, this is inefficient af. The question is not whether but how much it slows pathing down :/
+    @Override
+    public boolean isCloseToHome(int x, int y, int z) {
+        try {
+            IWaypoint[] waypoints = ForWaypoints.getWaypointsByTag(baritone, IWaypoint.Tag.HOME);
+            if (waypoints.length > 0) {
+                for (IWaypoint wp : waypoints) {
+                    boolean isClose = wp.getLocation().getDistance(x, y, z) <= Baritone.settings().homeProtectionRange.value;
+                    if (isClose) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
```

Comment:
Make this a method of `CalculationContext` and cache the waypoint *positions* as well as `homeProtectionRangeSq` (sqared radius) in the constructor, just like it is done for settings and other things that can change while pathing.
Then compare the summed squares to the squared distance instead of taking the square root every time (it's hidden in `wp.getLocation().getDistance(x, y, z)`)

Reply:
Yeah, what zacsharp said is all correct.

Make local fields of `CalculationContext`. Something like `boolean homeProtectionEnabled` and `int homeProtectionRange`. I could be convinced of just `int homeProtectionRange` and `if (homeProtectionRange == -1) {` everywhere, but since this is **such** a core inner function, I would Feel Safer if the entire thing was shoved under a nice little `if (homeProtectionEnabled) {` in `CalculationContext.costOfPlacingAt` and `CalculationContext.breakCostMultiplierAt`. Then, when that boolean is false, it's nothing but a never-taken branch, which is basically free.

It might be interesting to do something similar to the `Favoring` system for avoiding mobs where it generates spheres and stores them in a fastutil hashset. It would depend on how big the radius is and how many home waypoints you have. Gut feeling is that this is correct like this and you shouldn't do the `Favoring` approach.